cogboard-app/src/main/kotlin/com/cognifide/cogboard/widget/type/ZabbixWidget.kt

```diff
@@ -0,0 +1,209 @@
+    }
+
+    private fun attachCredentialsToBody(): JsonObject {
+        val credentials = JsonObject()
+                .put(CogboardConstants.PROP_USER, user)
+                .put(CogboardConstants.PROP_PASSWORD, password)
+        return prepareRequestBody(USER_LOGIN, credentials)
+    }
+
+    override fun handleResponse(responseBody: JsonObject) {
+        val result = extractResult(responseBody)
+        when {
+            hasResponseData(result) -> sendUpdate(result)
+            isAuthorized(result) -> fetchItemData()
+            hasError(responseBody) -> sendAuthorizationError(responseBody)
+            else -> {
+                saveToken(result.toString())
+                fetchItemData()
+            }
+        }
+    }
+
+    private fun extractResult(responseBody: JsonObject): Any {
+        val body = responseBody.getString(BODY)
+        val value = JsonObject(body)
+        return value.getValue(RESULT, "")
+    }
+
+    private fun hasResponseData(result: Any) = result.toString().contains("key_")
+
+    private fun sendUpdate(result: Any) {
+        val responseParams = result as JsonArray
+        val lastValue = responseParams.extractValue(LAST_VALUE)
+        send(JsonObject()
+                .put(LAST_VALUE, lastValue)
+                .put(NAME, responseParams.extractValue(NAME))
+                .put(HISTORY, modifyHistory(responseParams))
+                .put(CogboardConstants.PROP_WIDGET_STATUS, getStatusResponse(lastValue)))
+    }
+
+    private fun modifyHistory(responseParams: JsonArray): JsonObject {
+        val history = fetchHistoryFromContent()
+        val lastUpdate = responseParams.extractValue(LAST_CLOCK).toLong()
+        val filteredHistory = history.filter { it.key.toLong() > lastUpdate.minus(HOUR_IN_MILLS) }
+        val lastValue = responseParams.extractValue(LAST_VALUE)
+        return JsonObject(filteredHistory).mergeIn(JsonObject().put(lastUpdate.toString(), lastValue), true)
+    }
+
+    private fun fetchHistoryFromContent(): Map<String, Any> {
+        val content = boardServ.getContent(config.getString(CogboardConstants.PROP_ID))
+        return content.getJsonObject(HISTORY, JsonObject()).map
+    }
+
+    private fun getStatusResponse(lastValue: String): Widget.Status {
+        val convertedValue = lastValue.toDouble()
+        return if (range.isEmpty) {
+            Widget.Status.UNKNOWN
+        } else {
+            if (maxValue == 0) {
+                status(convertedValue)
+            } else {
+                val percentageValue = convertedValue.div(maxValue * 10.0.pow(9)).times(100)
+                status(percentageValue)
+            }
+        }
+    }
+
+    private fun status(value: Double): Widget.Status {
+        val first = range.list[0] as Int
+        val second = range.list[1] as Int
+        return when {
+            value < first -> {
+                return Widget.Status.OK
+            }
+            first <= value && value <= second -> {
```

Comment:
use ranges, example:
```
    val someRange = 1..100
    println(20 in someRange) // true
    println(120 in someRange) // false
```

Reply:
'contains(Double): Boolean' is deprecated for a while

Reply:
Ho about something like this:
```
val unstableRange = (range.list[0] as Int)..(range.list[1] as Int)
return when {
            value < unstableRange.start -> {
                return Widget.Status.OK
            }
            value in unstableRange -> {
                return Widget.Status.UNSTABLE
            }
            value > unstableRange.endInclusive -> {
                return Widget.Status.FAIL
            }
            else -> Widget.Status.UNKNOWN
        }
```

Reply:
fixed